source/extensions/transport_sockets/starttls/starttls_socket.cc

```diff
@@ -0,0 +1,32 @@
+#include "extensions/transport_sockets/starttls/starttls_socket.h"
+
+#include <iostream>
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace StartTls {
+
+// Switch clear-text to secure transport.
+bool StartTlsSocket::startSecureTransport() {
+  if (!using_ssl_) {
+    ssl_socket_->setTransportSocketCallbacks(*callbacks_);
+    ssl_socket_->onConnected();
+    oper_socket_ = std::move(ssl_socket_);
```

Comment:
Can transport sockets internally buffer? Is this guaranteed to be correct? Can we drop buffered data if there is flow control?

Reply:
I think that in this case we are safe. RawBuffer::doWrite/doRead do not buffer. 
In general case, where any type of transport socket could be plugged in, it would be nice to have ::flush() method to call before switching.

Reply:
Please add a TODO/comment about this.

Reply:
Added TODO.